File: Source/matlab/method/PseudoRandomSignal.m
classdef PseudoRandomSignal < handle
    properties (GetAccess = public, SetAccess = private)
        frequencyList
        nSequence
    end

    properties (Access = private)
        sequence
    end

    methods (Access = public)
        function prs = PseudoRandomSignal(frequencyList, method)
            arguments
                frequencyList (1, :) double {mustBePositive, mustBeInteger} = 1
                method               string = 'default'
            end

            switch method
                case '2n'
                    assert(isscalar(frequencyList), ...
                        'For ''2n'' method, frequencyList should be a scalar.');
                    prs.frequencyList = 2.^((1:frequencyList)-1);
                    prs.MakeSequence();
                case 'pattern'
                    prs.sequence      = cat(2, frequencyList, -frequencyList);
                    prs.nSequence     = length(prs.sequence);
                    prs.frequencyList = 1;           
                otherwise
                    prs.frequencyList = frequencyList;
                    prs.MakeSequence();
            end

        end

        function signal = Sampling(prs, n_interpolation, sampling_time)
            arguments
                prs             (1,1) PseudoRandomSignal
                n_interpolation (1,1) double {mustBeInteger, mustBePositive} = 1
                sampling_time   (1,1) double {mustBeInteger, mustBePositive} = 1
            end
            signal = repmat(repelem(prs.sequence, n_interpolation), ...
                                    1, sampling_time);
        end
    end

    methods (Access = private)
        function prs = MakeSequence(prs)
            nFrequency         = length(prs.frequencyList);
            n_cols_half_matrix = prs.frequencyList(1);
            for i = 2:nFrequency
                n_cols_half_matrix = lcm(n_cols_half_matrix, prs.frequencyList(i));
            end
            n_cols_matrix  = n_cols_half_matrix*2;
            sequenceMatrix = zeros(nFrequency, n_cols_matrix);
            prs.sequence   = zeros(1, n_cols_matrix);
            prs.nSequence  = n_cols_matrix;
            for i = 1:nFrequency
                half_period          = ones(1, n_cols_half_matrix/prs.frequencyList(i));
                pattern              = cat(2, half_period, -half_period);
                sequenceMatrix(i, :) = repmat(pattern, 1, prs.frequencyList(i));
            end
            prs.sequence = sign(sum(sequenceMatrix, 1));
        end
    end

end

File: Source/test/时间域绘图子窗口测试/Method/PseudoRandomSignal.cs
// PseudoRandomSignal.cs
using System;
using System.Linq;
using MathNet.Numerics;

namespace 时间域绘图测试

{
    public class PseudoRandomSignal
    {
        public double[] FrequencyList { get; private set; }
        public int NSequence { get; private set; }

        private int[] Sequence { get; set; }

        public PseudoRandomSignal(double[] frequencyList)
        {
            FrequencyList = frequencyList;
            MakeSequence();
        }

        private void MakeSequence()
        {
            int nFrequency = FrequencyList.Length;

            // 找到所有频率的最小公倍数作为基础周期的一半
            int nColsHalfMatrix = (int)FrequencyList[0];
            for (int i = 1; i < nFrequency; i++)
            {
                nColsHalfMatrix = LeastCommonMultiple(nColsHalfMatrix, (int)FrequencyList[i]);
            }

            int nColsMatrix = nColsHalfMatrix * 2;
            int[,] sequenceMatrix = new int[nFrequency, nColsMatrix];
            Sequence = new int[nColsMatrix];
            NSequence = nColsMatrix;

            for (int i = 0; i < nFrequency; i++)
            {
                double freq = FrequencyList[i];
                int halfPeriodLength = nColsHalfMatrix / (int)freq;

                int[] halfPeriod = new int[halfPeriodLength];
                for (int k = 0; k < halfPeriodLength; k++) halfPeriod[k] = 1;

                int[] pattern = new int[halfPeriodLength * 2];
                Array.Copy(halfPeriod, 0, pattern, 0, halfPeriodLength);
                for (int k = 0; k < halfPeriodLength; k++) pattern[halfPeriodLength + k] = -1;

                int[] repeatedPattern = new int[pattern.Length * (int)freq];
                for (int rep = 0; rep < freq; rep++)
                {
                    Array.Copy(pattern, 0, repeatedPattern, rep * pattern.Length, pattern.Length);
                }

                for (int col = 0; col < nColsMatrix; col++)
                {
                    int idx = col % repeatedPattern.Length;
                    sequenceMatrix[i, col] = repeatedPattern[idx];
                }
            }

            // Sum across all frequencies and take sign
            for (int col = 0; col < nColsMatrix; col++)
            {
                double sum = 0.0;
                for (int row = 0; row < nFrequency; row++)
                {
                    sum += sequenceMatrix[row, col];
                }
                Sequence[col] = sum >= 0 ? 1 : -1;
            }
        }

        public int[] Sampling(int nInterpolation, int samplingTime)
        {
            int originalLength = Sequence.Length;
            int newLength = originalLength * nInterpolation * samplingTime;
            int[] result = new int[newLength];

            for (int i = 0; i < originalLength; i++)
            {
                int val = Sequence[i];
                for (int rep = 0; rep < nInterpolation * samplingTime; rep++)
                {
                    result[i * nInterpolation * samplingTime + rep] = val;
                }
            }

            return result;
        }

        private static int GreatestCommonDivisor(int a, int b)
        {
            while (b != 0)
            {
                int temp = b;
                b = a % b;
                a = temp;
            }
            return a;
        }

        private static int LeastCommonMultiple(int a, int b)
        {
            return a / GreatestCommonDivisor(a, b) * b;
        }
    }

}